File: src/main/scala/com/blogspot/nurkiewicz/cacheable/CachingCalculatorDecorator.scala
package com.blogspot.nurkiewicz.cacheable

import calculator.Calculator
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.beans.factory.annotation.Configurable
import collection.JavaConversions._


class CachingCalculatorDecorator(target: Calculator) extends Calculator {
	private val cache = new java.util.concurrent.ConcurrentHashMap[java.lang.Integer, java.lang.Integer]

	override def identity(x: Int) = cache.get(x) match {
		case null =>
			val v = target.identity(x)
			cache.put(x, v)
			v
		case v => v
	}
}

File: src/main/java/com/blogspot/nurkiewicz/cacheable/calculator/CachingCalculatorDecorator.java
package com.blogspot.nurkiewicz.cacheable.calculator;

import java.util.Map;

/**
 * @author Tomasz Nurkiewicz
 * @since 1/12/13, 4:59 PM
 */
public class CachingCalculatorDecorator implements Calculator {

	private final Map<Integer, Integer> cache = new java.util.concurrent.ConcurrentHashMap<Integer, Integer>();

	private final Calculator target;

	public CachingCalculatorDecorator(Calculator target) {
		this.target = target;
	}

	@Override
	public int identity(int x) {
		final Integer existing = cache.get(x);
		if (existing != null) {
			return existing;
		}
		final int newValue = target.identity(x);
		cache.put(x, newValue);
		return newValue;
	}
}
